### ISSUE TITLE
Support for unicode utf-8 charmaps

### ISSUE THREAD

**Issue by thecodelearner** [awdeorio/mailmerge#97]
Using mailmerge to send mass emails is better than using any gmail extensions, since i have the ability to use advanced markdown and html
however while working with a large (5k+) csv file of job applicants, i came across an issue that people had entered their names in a localized language (eg. mandarin, etc) that sometimes didn't fit the ascii charmaps.

It would be useful to open the csv file as `encoding=utf8` with `errors=ignore` to avoid errors like this:
`'charmap' codec can't decode byte 0x9d in position 1710: character maps to <undefined>`
instead of just returning true if strings are ascii.

**Comment by thecodelearner** [comment]
also I would like to contribute towards this but i am confused since i have not worked on any open source stuff on github, especially in python 😅

**Comment by awdeorio** [comment]
It sounds to me like perhaps there is a problem with the CSV file.  I'm hesitant to ignore errors because some users might genuinely have a problem with their CSV file.  Perhaps another solution is produce a better error message.

Could you post the *smallest* CSV file that creates this error?  Maybe just one line?

**Comment by thecodelearner** [comment]
Sure, here's a raw CSV which does throw an error

```
name,email
"大松 李",person@example.com
"Juán",juan@example.com
"Jöse Felix",jisefelix@example.com
"Paveł",pavel@example.com

```

So anything like these might throw up the UnicodeDecodeError 😅

**Comment by awdeorio** [comment]
Hmm, I'm having trouble replicating the problem using your database.  Perhaps you could try attaching the csv file to this issue instead of copy-paste?

```console
$ mailmerge --version
mailmerge, version 2.1.0
$ cat mailmerge_database.csv 
name,email
"大松 李",person@example.com
"Juán",juan@example.com
"Jöse Felix",jisefelix@example.com
"Paveł",pavel@example.com
$ file mailmerge_database.csv 
mailmerge_database.csv: UTF-8 Unicode text
$ mailmerge --no-limit
>>> message 1
TO: person@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 11:31:47 -0000

Hi 大松 李

>>> message 1 sent
>>> message 2
TO: juan@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 11:31:47 -0000

Hi Juán

>>> message 2 sent
>>> message 3
TO: jisefelix@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 11:31:47 -0000

Hi Jöse Felix

>>> message 3 sent
>>> message 4
TO: pavel@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 11:31:47 -0000

Hi Paveł

>>> message 4 sent
>>> This was a dry run.  To send messages, use the --no-dry-run option.
```

**Comment by thecodelearner** [comment]
Seems like it must be an issue with my file.
Thanks for your help sire.
I'll test once again and if there's no issues, it would be better to close this issue

**Comment by thecodelearner** [comment]
GitHub doesn't support csv uploads so here's a txt file of the csv:
[mailmerge_database_txt.txt](https://github.com/awdeorio/mailmerge/files/4919320/mailmerge_database_txt.txt)

also, I think that my python installation on windows is broken since the same file runs perfectly on linux, but gives the UnicodeDecodeError on Windows

---> Running mailmerge on Windows:
```
PS7 mailmerge-client> mailmerge
Traceback (most recent call last):
  File "c:\users\pc\appdata\local\programs\python\python38\lib\runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\users\pc\appdata\local\programs\python\python38\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\PC\AppData\Local\Programs\Python\Python38\Scripts\mailmerge.exe\__main__.py", line 7, in <module>
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\click\core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\click\core.py", line 782, in main
    rv = self.invoke(ctx)
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\click\core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\mailmerge\__main__.py", line 114, in main
    for _, row in enumerate_range(csv_database, start, stop):
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\mailmerge\__main__.py", line 288, in enumerate_range
    for i, value in enumerate(iterable):
  File "c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\mailmerge\__main__.py", line 273, in read_csv_database
    for row in reader:
  File "c:\users\pc\appdata\local\programs\python\python38\lib\csv.py", line 110, in __next__
    self.fieldnames
  File "c:\users\pc\appdata\local\programs\python\python38\lib\csv.py", line 97, in fieldnames
    self._fieldnames = next(self.reader)
  File "c:\users\pc\appdata\local\programs\python\python38\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 17: character maps to <undefined>
```


---> Running mailmerge on Linux with exact same database.csv:

```
root@Proton:/mnt/c/Users/PC/merge-test# cat mailmerge_database.csv
name,email
"大松 李",person@example.com
"Juán",juan@example.com
"Jöse Felix",jisefelix@example.com
"Paveł",pavel@example.com
root@Proton:/mnt/c/Users/PC/merge-test# file mailmerge_database.csv
mailmerge_database.csv: CSV text
root@Proton:/mnt/c/Users/PC/merge-test# mailmerge --version
mailmerge, version 2.1.0
root@Proton:/mnt/c/Users/PC/merge-test# mailmerge
>>> message 1
TO: person@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 14:13:53 -0000

Hi, 大松 李

>>> message 1 sent
>>> Limit was 1 message.  To remove the limit, use the --no-limit option.
>>> This was a dry run.  To send messages, use the --no-dry-run option.
```

**Comment by seshrs** [comment]
@awdeorio I wonder if this is related to the behavior of the `open()` method in Python. According to the [docs](https://docs.python.org/3/library/functions.html#open), if no encoding is specified:

> The default encoding is platform dependent (whatever [locale.getpreferredencoding()](https://docs.python.org/3/library/locale.html#locale.getpreferredencoding) returns)

On my Mac, this returns UTF-8 but I think the Windows default is something else. In fact, it looks like Python is planning to output warnings if no encoding is specified to `open()` (see [PEP 597](https://www.python.org/dev/peps/pep-0597/)).

---

@thecodelearner do you want to try to verify if this fixes the issue for you? (It could make a great first-contribution! 😃 )

You'll need to ["fork" the repository](https://docs.github.com/en/github/getting-started-with-github/fork-a-repo#fork-an-example-repository), clone your fork locally and setup your development environment as described in [CONTRIBUTING.md](https://github.com/awdeorio/mailmerge/blob/develop/CONTRIBUTING.md). Then, find the spot where `mailmerge` opens the CSV file and change the encoding to UTF-8 as you had suggested.

Let us know if you have questions about any of this.

**Comment by thecodelearner** [comment]
@seshrs awesome! I will try this.

**Comment by thecodelearner** [comment]
Seems like some dependent modules are unix only, `pytest` fails at those.
As of opening the csv as UTF-8, I think it should work
As a safe measure here's the pytest log (running on windows, python 3.8 amd64)

```
(venv) PS7 mailmerge> pytest
=========================================================================== test session starts ===========================================================================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: E:\CodeH\Open-Source\mailmerge
plugins: cov-2.10.0
collected 61 items / 2 errors / 59 selected

(venv) PS7 mailmerge> python3
(venv) PS7 mailmerge> pytest --cov ./mailmerge
=========================================================================== test session starts ===========================================================================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: E:\CodeH\Open-Source\mailmerge
plugins: cov-2.10.0
collected 61 items / 2 errors / 59 selected

================================================================================= ERRORS ================================================================================== 
___________________________________________________________________ ERROR collecting tests/test_main.py ___________________________________________________________________ 
(venv) PS7 mailmerge> ls

    Directory: E:\CodeH\Open-Source\mailmerge

Mode                 LastWriteTime         Length Name
(venv) PS7 mailmerge> pytest --cov ./mailmerge
=========================================================================== test session starts ===========================================================================
platform win32 -- Python 3.8.3, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: E:\CodeH\Open-Source\mailmerge
plugins: cov-2.10.0
collected 61 items / 2 errors / 59 selected

================================================================================= ERRORS ==================================================================================
___________________________________________________________________ ERROR collecting tests/test_main.py ___________________________________________________________________
ImportError while importing test module 'E:\CodeH\Open-Source\mailmerge\tests\test_main.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests\test_main.py:13: in <module>
    import sh
c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\sh.py:34: in <module>
    raise ImportError("sh %s is currently only supported on linux and osx. \
E   ImportError: sh 1.13.1 is currently only supported on linux and osx. please install pbs 0.110 (http://pypi.python.org/pypi/pbs) for windows support.
_______________________________________________________________ ERROR collecting tests/test_main_output.py ________________________________________________________________
ImportError while importing test module 'E:\CodeH\Open-Source\mailmerge\tests\test_main_output.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
tests\test_main_output.py:14: in <module>
    import sh
c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\sh.py:34: in <module>
    raise ImportError("sh %s is currently only supported on linux and osx. \
E   ImportError: sh 1.13.1 is currently only supported on linux and osx. please install pbs 0.110 (http://pypi.python.org/pypi/pbs) for windows support.
============================================================================ warnings summary ============================================================================= 
c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\future\standard_library\__init__.py:65
  c:\users\pc\appdata\local\programs\python\python38\lib\site-packages\future\standard_library\__init__.py:65: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/latest/warnings.html

----------- coverage: platform win32, python 3.8.3-final-0 -----------
Name                            Stmts   Miss  Cover
---------------------------------------------------
mailmerge\__init__.py               3      0   100%
mailmerge\__main__.py             128     89    30%
mailmerge\exceptions.py             1      0   100%
mailmerge\sendmail_client.py       54     45    17%
mailmerge\template_message.py     105     81    23%
mailmerge\utils.py                 20     11    45%
---------------------------------------------------
TOTAL                             311    226    27%

========================================================================= short test summary info ========================================================================= 
ERROR tests/test_main.py
ERROR tests/test_main_output.py
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 
====================================================================== 1 warning, 2 errors in 0.78s ======================================================================= 
(venv) PS7 mailmerge>
```

**Comment by seshrs** [comment]
@thecodelearner Were you able to dry-run `mailmerge` on Windows with the updated code? (I believe it should still work even if the unit tests don't.)

(Could you also tell us what `locale.getprefferedencoding()` returns for you in Python on your Windows machine?)

**Comment by thecodelearner** [comment]
ah okay I'll post what `locale.getpreferedencoding()` returns when i log back into windows
for now, in linux there's a catch:

```
(env) $pip install --editable .[dev]        
zsh: no matches found: .[dev]
```

**Comment by seshrs** [comment]
Based on https://stackoverflow.com/a/30539963, I think you'll need to escape the square brackets on `zsh` or perhaps surround the argument in quotes.

**Comment by thecodelearner** [comment]
rookie mistake 😅

pytest report:

```
env/lib/python3.8/site-packages/future/standard_library/__init__.py:65
  /home/sahil/code/Open-Source/mailmerge/env/lib/python3.8/site-packages/future/standard_library/__init__.py:65: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

-- Docs: https://docs.pytest.org/en/latest/warnings.html

----------- coverage: platform linux, python 3.8.3-final-0 -----------
Name                            Stmts   Miss  Cover
---------------------------------------------------
mailmerge/__init__.py               3      0   100%
mailmerge/__main__.py             128     30    77%
mailmerge/exceptions.py             1      0   100%
mailmerge/sendmail_client.py       54      0   100%
mailmerge/template_message.py     105      1    99%
mailmerge/utils.py                 20      2    90%
---------------------------------------------------
TOTAL                             311     33    89%

```

will get back to Windows and post what `locale.getpreferedencoding()` returns and try a dry run for mailmerge.

**Comment by thecodelearner** [comment]
```>>> locale.getpreferredencoding()
'cp1252'
```

https://en.wikipedia.org/wiki/Windows-1252

**Comment by thecodelearner** [comment]
Yes, encoding="utf-8" solves the error on Windows:
here's a dry run

```
(env) PS7 mailmerge> mailmerge
>>> message 1
TO: person@example.com
SUBJECT: Testing mailmerge
FROM: My Self <myself@mydomain.com>
MIME-Version: 1.0
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: base64
Date: Tue, 14 Jul 2020 15:48:03 -0000

Hi, 大松 李!

>>> message 1 sent
>>> Limit was 1 message.  To remove the limit, use the --no-limit option.
>>> This was a dry run.  To send messages, use the --no-dry-run option.
```

**Comment by awdeorio** [comment]
Closed by #98
